src/gp_symbolic_regression.py

```diff
@@ -0,0 +1,162 @@
+    :param divisor: Value to divide by
+    :return: The quotient; the result of the division (dividend/divisor)
+    """
+    if divisor == 0:
+        return 999_999_999
+    else:
+        return dividend / divisor
+
+
+def mean_squared_error(
+    compiled_individual, independent_variables: list[list[float]], dependent_variable: list[float]
+) -> float:
+    """
+    Calculate the mean squared error for a given function on the observed data. The independent variables are a list of
+    lists where the list of the outer list is the number of observations and the length of the inside list is the number
+    of independent variables. The length of the dependent variables is the number of observations.
+
+    :param compiled_individual: A callable function that the dependent variables are give to as arguments
+    :param independent_variables: Values given to the function to predict
+    :param dependent_variable: Expected result of the function
+    :return: Mean squared error of the function on the observed data
+    """
+    running_error_squared_sum = 0
+    for xs, y in zip(independent_variables, dependent_variable):
+        y_hat = compiled_individual(*xs)
+        running_error_squared_sum += (y - y_hat) ** 2
+    return running_error_squared_sum / len(dependent_variable)
+
+
+def mean_squared_error_fitness(
+    individual, toolbox, independent_variables: list[list[float]], dependent_variable: list[float]
+) -> tuple[float]:
+    """
+    Calculate the mean squared error for a given function on the observed data. The independent variables are a list of
+    lists where the list of the outer list is the number of observations and the length of the inside list is the number
+    of independent variables. The length of the dependent variables is the number of observations.
+
+    :param individual: A deap representation of a function that the dependent variables are give to as arguments
+    :param toolbox: A deap toolbox
+    :param independent_variables: Values given to the function to predict
+    :param dependent_variable: Expected result of the function
+    :return: A tuple containing the mean squared error of the function on the observed data
+    :raises ValueError: If the number of observations s 0 or if there are a different number of observations
+    """
+    if len(independent_variables) == 0 or len(dependent_variable) == 0:
+        raise ValueError("Number of dependent and independent variables must be greater than 0")
+    if len(independent_variables) != len(dependent_variable):
+        raise ValueError(
+            f"Uneven number of observations of independent and dependent variables: "
+            f"{len(independent_variables)}, {len(dependent_variable)}"
+        )
+    callable_function = toolbox.compile(expr=individual)
+    individual_mean_squared_error = mean_squared_error(callable_function, independent_variables, dependent_variable)
+    return (individual_mean_squared_error,)
```

Comment:
why a tuple? is that just a DEAP thing?

Reply:
Yeah. It's for doing multi objectives easy